### DISCUSSION TITLE
Label: MaxLines and LineBreakMode

### DISCUSSION THREAD

**Discussion by Xandroid4Net** [dotnet/maui#5492] [Ideas]
I am trying to become more familiar with .net MAUI and the app I was writing I decided I wanted to have a label with at most 2 lines of text and then truncate at the end or TailTruncation.

After running the code it was not working, it always had one line with the tail truncated, so I decided to look at the source and it appears that when the linebreakmode is set to TailTruncation the maui framework overrides my setting of maxlines and sets maxlines to 1.

1. Am I reading the code right? I am looking at the SetLineBreakMode extension method. ([iOS](https://github.com/dotnet/maui/blob/3717ce0d385d1aadbd3b9795c203d438426aae17/src/Core/src/Platform/iOS/LabelExtensions.cs#L94), [Android](https://github.com/dotnet/maui/blob/3717ce0d385d1aadbd3b9795c203d438426aae17/src/Core/src/Platform/Android/TextViewExtensions.cs))
2. What is the reasoning behind the maxlines = 1 when using TailTruncation? I imagine there might be some other platform quirk that I don't know about.
3. Is there some other way that I can achieve tailtruncation with maxlines > 1?

**Comment by ram16g** [comment]
fixed in next release see  [ this pr](https://github.com/dotnet/maui/pull/5093)

**Comment by LennoxP90** [comment]
This is still an issue, If i have a label with TailTruncation and MaxLines at 2
![image](https://user-images.githubusercontent.com/1734719/191092732-5d80a993-4b2d-4565-bb0a-7ccc63ba46f1.png)

**Comment by hartez** [comment]
The behavior is for backward compatibility with Xamarin.Forms. If you want a custom (and more sensible) behavior in your .NET MAUI app, you can see how to modify it in [this repo](https://github.com/hartez/MultilineTruncate#readme). It's basically a few lines you cut-and-paste into your `MauiProgram.cs`.

**Comment by LennoxP90** [comment]
what about iOS?

**Reply by hartez** [reply]
> what about iOS?

I don't know, haven't looked into it. If multiline UILabels with tail truncation are possible on iOS, then I imagine you could do something similar to what I posted and make it work.

**Comment by needledrag** [comment]
Did you manage to find a workaround for this, having similar issues but looking for a multiplatform solution?

Cheers
Shaun

**Comment by LennoxP90** [comment]
Place this in MauiProgram.cs

```
static void AllowMultiLineTruncation()
  {
    static void UpdateMaxLines( Microsoft.Maui.Handlers.LabelHandler handler, ILabel label )
    {
#if ANDROID
      var textView = handler.PlatformView;
      if(    label is Label controlsLabel 
          && textView.Ellipsize == Android.Text.TextUtils.TruncateAt.End )
      {
        textView.SetMaxLines( controlsLabel.MaxLines );
      }
#elif IOS
      var textView = handler.PlatformView;
      if( label is Label controlsLabel
          && textView.LineBreakMode == UILineBreakMode.TailTruncation )
      {
        textView.Lines = controlsLabel.MaxLines;
      }  
#endif
    };

    Label.ControlsLabelMapper.AppendToMapping(
       nameof( Label.LineBreakMode ), UpdateMaxLines );

    Label.ControlsLabelMapper.AppendToMapping(
      nameof( Label.MaxLines ), UpdateMaxLines ); 
  }
```
Call above code in your CreateMauiApp() function

**Reply by NehpetsEtnam** [reply]
Thanks for this work around. But I think we need to add this `&& controlsLabel.MaxLines != -1`  to the if statement coz if not and your Label has `LineBreakMode="TailTruncation"` without `MaxLines` being set, it will have a bigger space on top and bottom. Tested it on Pixel 4 API 32. 

Without `&& controlsLabel.MaxLines != -1`
<img width="354" alt="Screen Shot 2023-01-21 at 9 35 39 AM" src="https://user-images.githubusercontent.com/33170854/213832372-d52cb4dc-8a81-4997-ad3a-25bbbe7bc2d4.png">

With `&& controlsLabel.MaxLines != -1`
<img width="351" alt="Screen Shot 2023-01-21 at 9 34 27 AM" src="https://user-images.githubusercontent.com/33170854/213832346-79e93e4d-98ee-4bb4-9e0e-94c131422494.png">

**Comment by needledrag** [comment]
Wow, that was quick, many thanks, I'll give this a go.

Cheers
Shaun

**Comment by Daniel0b1b** [comment]
Thanks for the workaround!

**Comment by hansmbakker** [comment]
Is this not solved by https://github.com/dotnet/maui/pull/14918?

**Reply by acaliaro** [reply]
I have the problem on ios with a label on a grid inside a border

**Comment by vchelaru** [comment]
I have updated to Maui 9.0 and I no longer have `Label.ControlsLabelMapper`. Is this expected?
